Ensure k8s apply works with check mode

willthames opened the pull request:
##### SUMMARY

Update the new predicted object with fields from the previous object
before applying in check mode

Don't log output of `file` with `state: absent` on huge virtualenvs!

Fixes #60510

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
module_utils/k8s/raw.py

willthames committed:
Ensure k8s apply works with check mode

Update the new predicted object with fields from the previous object
before applying in check mode

Don't log output of `file` with `state: absent` on huge virtualenvs!

Fixes #60510

fabianvf commented:
shipit

mitchellmaler commented:
I tested your repo/branch and some of the differences did clean up and not show any changes. I did however get some changes like this and this is with the default apply enabled.

```yaml
--- before
+++ after
@@ -33,7 +33,6 @@
         - secretRef:
             name: newrelic-license
         image: newrelic/infrastructure-k8s:1.9.2
-        imagePullPolicy: IfNotPresent
         name: newrelic-infra
         resources:
           limits:
@@ -45,9 +44,6 @@
           capabilities:
             add:
             - SYS_PTRACE
-          procMount: Default
-        terminationMessagePath: /dev/termination-log
-        terminationMessagePolicy: File
         volumeMounts:
         - mountPath: /host
           name: host-volume
```

willthames commented:
ok, there is more of a bug with `check_mode` for `apply` than I realised

@fabianvf I've raised https://github.com/openshift/openshift-restclient-python/pull/315 for the openshift end and I have the ansible fix ready to go once openshift has been released with the fix.

mitchellmaler commented:
Doing a bit more testing if the object it checks against doesn't exist it will error with this
` AttributeError: 'NoneType' object has no attribute 'to_dict'`


```
ok: [localhost] => (item=servicemonitor-calico-node.yml)
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'NoneType' object has no attribute 'to_dict'
failed: [localhost] (item=servicemonitor-fluent-bit.yml) => changed=false 
  ansible_loop_var: item
  item: /builds/ansible/ansible_remote_roles/kubernetes_monitoring_config/files/servicemonitors/common/servicemonitor-fluent-bit.yml
  module_stderr: |-
    Traceback (most recent call last):
      File "/root/.ansible/tmp/ansible-tmp-1565976497.6722012-279090932748304/AnsiballZ_k8s.py", line 125, in <module>
        _ansiballz_main()
      File "/root/.ansible/tmp/ansible-tmp-1565976497.6722012-279090932748304/AnsiballZ_k8s.py", line 117, in _ansiballz_main
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
      File "/root/.ansible/tmp/ansible-tmp-1565976497.6722012-279090932748304/AnsiballZ_k8s.py", line 51, in invoke_module
        spec.loader.exec_module(module)
      File "<frozen importlib._bootstrap_external>", line 728, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/tmp/ansible_k8s_payload_tws7_oli/__main__.py", line 280, in <module>
      File "/tmp/ansible_k8s_payload_tws7_oli/__main__.py", line 276, in main
      File "/tmp/ansible_k8s_payload_tws7_oli/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py", line 182, in execute_module
      File "/tmp/ansible_k8s_payload_tws7_oli/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py", line 287, in perform_action
    AttributeError: 'NoneType' object has no attribute 'to_dict'
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1
```

willthames commented:
I've made this a WIP until the openshift changes are merged and released.

willthames commented:
Probably a 2.9 candidate backport - not sure there's much point having a changelog for a fix to a feature that only exists in 2.9 though

willthames committed:
Use openshift client fix to improve apply for check mode

Use new apply_object method to get a better approximation
of the expected object in check mode.

Requires released upgrade to openshift

willthames committed:
Add changelog fragment for k8s apply check mode fix

felixfontein commented on changelogs/fragments/60510-k8s-apply-check-mode.yml:
```diff
@@ -0,0 +1,2 @@
+minor_bugfixes:
```
```suggestion
bugfixes:
```
There's `bugfixes` and `minor_changes`, but no `minor_bugfixes`.

willthames committed:
Update changelogs/fragments/60510-k8s-apply-check-mode.yml

Co-Authored-By: Felix Fontein <felix@fontein.de>

felixfontein commented:
@willthames thanks for fixing this!
@fabianvf @mitchellmaler thanks for testing and reviewing!
